mlflow/sklearn/utils.py

```diff
@@ -97,21 +127,178 @@ def _truncate_dict(d, max_key_length=None, max_value_length=None):
         should_truncate_key = (not key_is_none) and (len(str(k)) > max_key_length)
         should_truncate_val = (not val_is_none) and (len(str(v)) > max_value_length)
 
-        new_k = str(k)[:max_key_length] if should_truncate_key else k
+        new_k = _truncate_and_ellipsize(k, max_key_length) if should_truncate_key else k
         if should_truncate_key:
-            msg = "Truncated the key `{}`".format(k)
+            # Use the truncated key for warning logs to avoid noisy printing to stdout
+            msg = "Truncated the key `{}`".format(new_k)
             _logger.warning(msg)
 
-        new_v = str(v)[:max_value_length] if should_truncate_val else v
+        new_v = _truncate_and_ellipsize(v, max_value_length) if should_truncate_val else v
         if should_truncate_val:
-            msg = "Truncated the value `{}` (in the key `{}`)".format(v, k)
+            # Use the truncated key and value for warning logs to avoid noisy printing to stdout
+            msg = "Truncated the value of the key `{}`. Truncated value: `{}`".format(new_k, new_v)
             _logger.warning(msg)
 
         truncated[new_k] = new_v
 
     return truncated
 
 
+def _get_meta_estimators_for_autologging():
+    """
+    :return: A list of meta estimator class definitions
+             (e.g., `sklearn.model_selection.GridSearchCV`) that should be included
+             when patching training functions for autologging
+    """
+    from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+    from sklearn.pipeline import Pipeline
+
+    return [
+        GridSearchCV,
+        RandomizedSearchCV,
+        Pipeline,
+    ]
+
+
+def _is_parameter_search_estimator(estimator):
+    """
+    :return: `True` if the specified scikit-learn estimator is a parameter search estimator,
+             such as `GridSearchCV`. `False` otherwise.
+    """
+    from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
+
+    parameter_search_estimators = [
+        GridSearchCV,
+        RandomizedSearchCV,
+    ]
+
+    return any(
+        [
+            isinstance(estimator, param_search_estimator)
+            for param_search_estimator in parameter_search_estimators
+        ]
+    )
+
+
+def _log_parameter_search_results_as_artifact(cv_results_df, run_id):
+    """
+    Records a collection of parameter search results as an MLflow artifact
+    for the specified run.
+
+    :param cv_results_df: A Pandas DataFrame containing the results of a parameter search
+                          training session, which may be obtained by parsing the `cv_results_`
+                          attribute of a trained parameter search estimator such as
+                          `GridSearchCV`.
+    :param run_id: The ID of the MLflow Run to which the artifact should be recorded.
+    """
+    with TempDir() as t:
```

Comment:
We could replace this with `tempfile.TemporaryDicretory`: https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

Reply:
I think the reason that we have been using this is that Python 2 (which we no longer support) doesn't support this context manager.

Reply:
I think `mlflow.utils.file_utils.TempDir()` provides a couple of nice additions to `TemporaryDirectory()`:

1. The `path()` function can be easily used to obtain the path to a location within the directory
2. The `chdr` option of the constructor allows callers to change the current working directory within the TempDir() context.

Given the utility of these extra features and the fact that `TempDir()` seems to have worked reliably throughout MLflow for the last couple of years, I'm inclined to continue using it. Let me know if you still feel strongly that we should leverage TemporaryDirectory instead, though!